src/common/meta/src/maintenance/reconcile_table/utils.rs

```diff
@@ -0,0 +1,705 @@
+                        region_metadata.column_metadatas.iter().map(ColumnMetadataDisplay).collect::<Vec<_>>()
+                    )
+                }
+            );
+            region_ids.push(region_metadata.region_id);
+        }
+    }
+
+    // TODO(weny): verify the new column metadatas are acceptable for regions.
+    Ok((latest_region_metadata.column_metadatas.clone(), region_ids))
+}
+
+/// Constructs a vector of [`ColumnMetadata`] from the provided table information.
+///
+/// This function maps each [`ColumnSchema`] to its corresponding [`ColumnMetadata`] by
+/// determining the semantic type (Tag, Timestamp, or Field) and retrieving the column ID
+/// from the `name_to_ids` mapping.
+///
+/// Returns an error if any column name is missing in the mapping.
+pub(crate) fn build_column_metadata_from_table_info(
+    column_schemas: &[ColumnSchema],
+    primary_key_indexes: &[usize],
+    name_to_ids: &HashMap<String, u32>,
+) -> Result<Vec<ColumnMetadata>> {
+    let primary_names = primary_key_indexes
+        .iter()
+        .map(|i| column_schemas[*i].name.as_str())
+        .collect::<HashSet<_>>();
+
+    column_schemas
+        .iter()
+        .map(|column_schema| {
+            let column_id =
+                *name_to_ids
+                    .get(column_schema.name.as_str())
+                    .context(UnexpectedSnafu {
+                        err_msg: format!(
+                            "Column name {} not found in name_to_ids",
+                            column_schema.name
+                        ),
+                    })?;
+
+            let semantic_type = if primary_names.contains(&column_schema.name.as_str()) {
+                SemanticType::Tag
+            } else if column_schema.is_time_index() {
+                SemanticType::Timestamp
+            } else {
+                SemanticType::Field
+            };
+            Ok(ColumnMetadata {
+                column_schema: column_schema.clone(),
+                semantic_type,
+                column_id,
+            })
+        })
+        .collect::<Result<Vec<_>>>()
+}
+
+/// Checks whether the schema invariants hold between the existing and new column metadata.
+///
+/// Invariants:
+/// - Primary key (Tag) columns must exist in the new metadata, with identical name and ID.
+/// - Timestamp column must remain exactly the same in name and ID.
+pub(crate) fn check_column_metadata_invariants(
+    new_column_metadatas: &[ColumnMetadata],
+    column_metadatas: &[ColumnMetadata],
+) -> bool {
+    let new_primary_keys = new_column_metadatas
+        .iter()
+        .filter(|c| c.semantic_type == SemanticType::Tag)
+        .map(|c| (c.column_schema.name.as_str(), c.column_id))
+        .collect::<HashMap<_, _>>();
+
+    let old_primary_keys = column_metadatas
+        .iter()
+        .filter(|c| c.semantic_type == SemanticType::Tag)
+        .map(|c| (c.column_schema.name.as_str(), c.column_id))
+        .collect::<HashMap<_, _>>();
```

Comment:
There's no need to build a hashmap; I see you just want to iterate over the old PKs.